### ISSUE TITLE
PP-OCRv5 mobile: spaces lost after 2.0.4 upgrade, isolated to resize / batch-width policy (2.2.0)

### ISSUE THREAD

**Issue by dusdong** [FarhanLodi/PaddleOcrNet#8]
## Summary

Upgrading a fixed PP-OCRv5 mobile / SLANetPlus setup from PaddleOcrNet 2.0.4 to 2.2.0 loses one U+0020 in each of four previously correct cells in a synthetic mixed Chinese/English table. The table topology remains 7 rows, 4 columns, 20 cells. I isolated a regression in the recognition input path, independently of table HTML/export or application string processing.

This is a recognition-quality / compatibility report, **not an assertion that CTC incorrectly removes ordinary spaces**. The changed preprocessing already appears in the 2.1.0 binary; I am reporting that the issue persists in 2.2.0, not that 2.2.0 first introduced it. I understand the newer preprocessing intentionally follows the Python pipeline, so a universal revert is not implied.

## Reproduction environment and fixed assets

- macOS 27.0, ARM64 CPU; .NET 10.0.11; SDK 10.0.401.
- PaddleOcrNet 2.2.0, EasyImageSharp 1.1.0, Microsoft.ML.OnnxRuntime 1.30.0.
- CPU: IntraOpNumThreads=2, InterOpNumThreads=1, ORT_ENABLE_ALL.
- `PP-OCRv5_mobile_rec.onnx`: SHA-256 `d253c3cbee6e507828a5271a30ab0ec8ae7c2a99d0cc8e6f844fe380809d22b3`.
- `ppocrv5_dict.txt`: SHA-256 `9dfc80c50b6cb07399a47a7cf25d11db475fb4ad0e1fc96b2eff6467c8166ff3`.
- These are the assets registered by [PaddleModelRegistry at the 2.2.0 release](https://github.com/FarhanLodi/PaddleOcrNet/blob/5a8bfe852b1a1b243a7145408c1de687db6793c1/src/PaddleOcrNet/Internal/PaddleModelRegistry.cs). The repro accepts existing files and downloads nothing.

## Causal isolation

1. Taking each version's returned quadrilaterals, using that version's actual `PerspectiveWarp.Rectify`, and invoking its actual internal recognizer reproduces all 20 original table-line strings, including the four changed spaces, on each side. No provider, HTML parser or export code is involved in this replay.
2. Keeping the newer crop pixels fixed and invoking the old 2.0.4 recognizer restores three of the four protected strings. Running that old recognizer with the newer EasyImageSharp/ORT dependencies preserves all 20 strings from its old-dependency run (14 confidences change). Thus the SDK code path has an effect independently of the dependency-version change.
3. The actual 2.0.4 binary uses Bicubic resize and a batch tensor width equal to the maximum `ceil(H*w/h)`; the newer recognizer uses Triangle resize and the minimum-320 / floor / cap policy in [`ComputeBatchTensorWidth` and `WriteNormalized`](https://github.com/FarhanLodi/PaddleOcrNet/blob/5a8bfe852b1a1b243a7145408c1de687db6793c1/src/PaddleOcrNet/Internal/Recognition/SvtrRecognizer.cs#L269).
4. With fixed newer crops, newer image library, ORT 1.30.0, the same model, grouping and decoder, changing only those preprocessing factors gives this matrix. "Old width policy" includes the prior width rounding, not just a padding constant. The cells are identified by zero-based row/column:

| Recognition resize / width policy | (0,0) | (2,2) | (6,0) | (6,1) |
|---|---|---|---|---|
| Triangle / new | space lost | space lost | space lost | space lost |
| Triangle / old | space lost | preserved | preserved | space lost |
| Bicubic / new | space lost | preserved | space lost | preserved |
| Bicubic / old | preserved | preserved | preserved | space lost |

There are interactions: no single tested setting preserves all four. Restoring old preprocessing globally is therefore **not** proposed as a safe fix. This particular fixture is grayscale; RGB/BGR tensor bytes are identical, so channel order does not explain these cases.

## Minimal two-forward reproduction

The following standalone console app embeds **one generated, non-customer RGB24 crop** (180x34) reading `结算 Pay 21190`. It calls the unmodified 2.2.0 DLL's normalization and CTC methods by reflection, with the same model session. It checks that effective normalized pixels are byte-identical and that padding is zero.

Only the tensor's total width changes: 279 (the legacy width of its original six-crop batch) versus 320 (the newer minimum). The effective resized width stays 255, input height 48, batch size 1. Thus this reduced test removes detection, crop choice, batch companions, concurrency, table extraction and decoder differences.

Observed twice in separate processes, with identical output:

```text
W=279, effective=255, text=[结算 Pay 21190], step10.argmax=18384, p(space)=0.59423226, p(blank)=0.40551534
W=320, effective=255, text=[结算Pay 21190], step10.argmax=0, p(space)=0.48934844, p(blank)=0.51035815
```

The space becomes a blank **in the model output at a shared timestep**, before CTC emits text. This supports a width-sensitive inference effect, not trimming in the result layer. I have not localized the responsible layer inside the ONNX graph.

Save the following two files in an empty directory, then run with the two matching existing model assets:

```sh
dotnet run -c Release -- /path/to/PP-OCRv5_mobile_rec.onnx /path/to/ppocrv5_dict.txt
```

Could this fixture be added to recognition regression coverage, and could the upgrade path offer or document a mitigation for models/crops affected by the changed preprocessing? The above is a narrow reproducible quality regression, not a claim that one preprocessing policy is globally more accurate.

<details>
<summary>PublicRepro.csproj</summary>

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup><OutputType>Exe</OutputType><TargetFramework>net10.0</TargetFramework><ImplicitUsings>enable</ImplicitUsings><Nullable>enable</Nullable></PropertyGroup>
  <ItemGroup><PackageReference Include="PaddleOcrNet" Version="[2.2.0]" /></ItemGroup>
</Project>
```
</details>

<details>
<summary>Program.cs (complete embedded synthetic fixture)</summary>

```csharp
using System.IO.Compression;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using EasyImageSharp;
using EasyImageSharp.PixelFormats;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using PaddleOcrNet.Models;

static class Program
{
    // Generated synthetic crop, RGB24, 180 x 34, reading "结算 Pay 21190".
    // No customer image/data and no font installation needed to reproduce.
    const string CropGzipBase64 = "H4sIAAAAAAAC/+2bd0wV2xbGsSsiVuxeG6hYKIK9PaIB/7D3biwRe732bowNUbFijx0VuyIGS4wVeMYGKokd1Niw9/Z+YeVO5p2ZM5xD1dzz/UFm9t4zs2fvb631rXWGX79ssMEGG2ywwQYbMg3x8fHfv38310vX7Nmzg4KCkr1PaGjonj17Pnz4oNsbHBy8f//+L1++2Bb8D4K3t3e5cuUiIyN1ez9//mxnZ1e6dGnjm7x48cLR0dHBweHly5fa3piYmKxZs3KTb9++2Rb8T8GFCxfY+vz5879+/To13BgzZgzDKlWq9LcGT5486dSpE70eHh5/6+H48eO2jfgN0a1bN3Zt0qRJ5gZYwo3Y2NjcuXPbmcGSJUuyZMliZx7ELIOb02syvkiRIjVq1GjduvW+ffsMomG64siRI61atWJZ8ubNW7NmTX9/f0zA3ODnz58TUjEcJj9y5EiD21o4kmGDBw92c3PLkyePs7MzhonfNjd49+7dDRs2zJcvX8GCBf38/E6dOmXJC/I6OXLksLe3f/bsmdKIddtZAGXmiYmJFStWpOXgwYM/f/6sXr0690xISJDeuLg4ppQ9e3YcFKe3b98m7rC5uqHHgBs5c+YslwS2gwVRpsEyPn36NCNZgaDq2LGjPJ134e3kmIMTJ05ox8MctWkY7LiFI4n+JUuWZAD2WLlyZYI1x0WLFr1+/bp28MSJE5XpQWMOGL9mzZpkX3P06NEMxqvLKTvL386dO3uoADkZw3Z7/D8CAgJEqTZr1owBbdq0kZvs2LGD0/Hjx3P848cPDJzTWbNmKQ9dsWIFLRiaVdzgiepGqIV9sSB01a9f/+vXrxnGjebNm/PQChUqREREKG6zTp06NJYtW1YrxZs2bfqfJBQuXNiYG5aMRMyLV+nbty8unZZXr161b99efPu7d+/Ug8PCwmiHbxs2bEDpffr0iY2Q3dQlkoIHDx7kypWL5RWl8ebNG09Pz8OHD1sVU6DTnDlzYC/e49atWygH5lOmTJkZM2YcT8KyZcuqVq0aHh5+/B8cPXoU8587dy7HkCdl3BBcvHhRDE2cUsYA59CgQQOTCILjxYcwEwOTbNGiRbIxJdmRS5cupYslVS8dxxgI7dOnT1cPrlWrFo3jxo1TN4rTwwMYTEDItnHjRrXzqVu3rngPq/SGZKZDhw61sxLC/BRzA2C/9C5fvjzTlZus+bBhw9KVG+wpXYGBgSbtSC/asUqFM3fv3qUlW7ZsasEAoqOjaScuf/z4UffpISEhDMATwgT8RlRUFMFL19VYmKco3NB6Hl0UL148TbjRpEkTeseOHatuxLUuXrzYx8enWLFiBFleE+arnzVz5kyuQibp3hOVS+/kyZOt4oZkamxrunKjWrVquotMyECS0XXz5k1pWbt2LadoGK2rF42kmx4STRCH9KJdCSuKFU+bNi3FOWxm+Y0SJUrQu2nTJqUlKCgIrSv3h+3ypsDb25uAK2MePnwoEk5b1UEwiNa9evWqVdzo06cPV/Xq1SsDuHHo0CFzXdu3b5fTUaNGcdq7d2/tSNIWA2fr4uKCtyHuY1wcM9LV1VVCw5UrVyzZVjIOXW5gsP+1ACK3UsmNPXv2yGTU7o41wVdg9ew798dzwhwxAWSwiaQcPny47j3Re9bGFEQXFxoUkNOEG2h+uhYuXGhux1l/OW3Xrp057yeFCyUHMcG9e/dkX/Ah+fPnx77QdUrW6WEBML1MjCnE0Hnz5jk6OtKFwaq70O3aUsOIESMY6eXlpc76aSHomFRIevbsSfuUKVOsIgZBmavwUffv309XbqD86cKcTSrMaHsxE7IAafH19eWUVdLeZMCAAXQNHDjQYA7oFonXWm2TAqRrTCGYOiehfPny9vb2yh0wfyVSGJeqGFyoUCG1fpbQc+zYMXXUllh87do1q969ZcuWXNW/f/+U7bjlI5FSYlYkvKhKIiB+vkuXLjjGv/76S02Gxo0bc7po0SLtTRDM5sKNybIjvdJEqKdrTDGpi+IBcIznz583vhb3gh8LCAiQpAyo6w9S3lEvUUREBC1EB6te/ODBg6JtDJxGWnEDnD59GvmnXhDYQhbv5+fH8apVq2SYnOJPzPmNQYMGmZsAekYkR2JiYnx8fFhY2Pz587mKrs2bN7cxxLZt236TPEUXb9++hQ+enp4Kl6pWrSrH6p+NYmNjRZAr2dyQIUNomTp1quXPevTokZTgkr0qrbghcRPtRFDo2rUr2kMK5u7u7lyFXlLXKEjQtJfjZ7TJnQI2V9R4qVKllKqvREwCzYQJE4xjgW44zpQ8RQuciSQvDRo0WLdunVTv8VRaboB69erRuHPnTjllNUyUrTGYP08RD59sHS8NuaEF0VD09p07d6QFqclpjx49tIOlULZy5UrdYqbUXSWOs+z4VehH5H38+DEDhBvh4eHaa7du3WqOGzdu3AhPArdiDLI5XIMcSVBOU1kX1QILkmhlkqCZ4wbkobFt27Yck7RyXKVKFQu3g2WUeheMMqkvZTw3CDRc4uTkpJQuN2zYQIubm5tWZIqGN/ejG34jODgYGaP7S0TKuKHgzJkzjOnQoYO2K3cSrIrmVnFj9erVugtijhtIOwcHB6b05s2bBQsWmCvy6BJDojZZHje35JJ05YZUV3Dd6iqW/LJG1DPxq7ST5lvitNWQhM4qbuzatcvkewz5WgMD1H6qkS0JJo1K3T713JDfkpDoJu3IdV1ugH79+tGOxCIucBATE2MJMfz9/WWFz507Z+HaphU3tBl6ZGQkq4oeMJm8REyTGo6kVN27d7dkzkSokJAQRDtRgPSQlNAqbpC42aUOym+4qefG3r17RTUdOHBA/CeRQqpA5rghdsS7E69dXV0tiewEcfmJ/OTJk5bbXZpwg6yBENaqVSuybN6O19m+fbsk41ptiUiQt0a44iVwkvKrGUICAWBOQe3YsQMC+Pr6kvKrtwkd8vz589TEFPIUOIy+1f3aJL1jCiGydu3a8i6kDzh8OYD88vWR7udtUEIuMfkd0+A3SvkiguUi4cKgcqugrimxwqX/gUyAEKa0REVFWTsSGeDs7CwTwGtJ5V80p654Y6eUT01ErHLJ+vXrDdQ12yeXkK00atRo/PjxGBqsUOsNPJI2e/Xy8jLHDTJB9hGbRW1iv7qPTm9uSH6HD3RxceFBLCPWJCyVX2x1uREYGCirQVZrYfZtAHUFzPhDqbNnz6ZgpJg/fGbvChQoQCgMDQ01mDC9JFNwg8EwMNnvvlgx/MylS5d0P+61NoeNjo7mEknzIfnRo0fNPTcF3MgAbNmyRT6K+GVDcrAqpuCFxI0TBxH579+/N7jz78kNkWe6JUQbTBAXFwcxlBCjBj4ZMWzyiWZCQgK+wpLPetcm4bd62cuXLxOCYazB57g2/KsghR3Uvkg7C8saNvwbULduXScnJ5FPPj4+tn+qskEp8bm7u5MDenp6kgdl1j+52GCDDTb8ofgfswAW77hHAAA=";
    delegate (string Text,float Confidence) Decode(ReadOnlySpan<float> logits,int time,int classes,IReadOnlyList<string> vocab,bool[]? mask);
    static void CheckHash(string file,string expected) {
        if(Convert.ToHexStringLower(SHA256.HashData(File.ReadAllBytes(file)))!=expected)throw new InvalidDataException("Unexpected asset: "+Path.GetFileName(file));
    }
    static void Main(string[] args) {
        if(args.Length!=2)throw new ArgumentException("Pass paths to PP-OCRv5_mobile_rec.onnx and ppocrv5_dict.txt");
        CheckHash(args[0],"d253c3cbee6e507828a5271a30ab0ec8ae7c2a99d0cc8e6f844fe380809d22b3");
        CheckHash(args[1],"9dfc80c50b6cb07399a47a7cf25d11db475fb4ad0e1fc96b2eff6467c8166ff3");
        using var packed=new MemoryStream(Convert.FromBase64String(CropGzipBase64));
        using var gzip=new GZipStream(packed,CompressionMode.Decompress);
        byte[] pixels=new byte[180*34*3];gzip.ReadExactly(pixels);
        if(gzip.ReadByte()!=-1||Convert.ToHexStringLower(SHA256.HashData(pixels))!="86ff7735eb51daf578d40c2520f4b96205cd7ef1a78b23d682bcb3207aa85867")throw new InvalidDataException("Fixture drift");
        using var crop=new Image<Rgb24>(180,34);
        crop.ProcessPixelRows(rows=>{for(int y=0;y<34;y++)pixels.AsSpan(y*180*3,180*3).CopyTo(MemoryMarshal.AsBytes(rows.GetRowSpan(y)));});
        using var options=new SessionOptions{IntraOpNumThreads=2,InterOpNumThreads=1,GraphOptimizationLevel=GraphOptimizationLevel.ORT_ENABLE_ALL};
        using var session=new InferenceSession(args[0],options);
        var sdk=typeof(RecognitionOptions).Assembly;
        var dict=sdk.GetType("PaddleOcrNet.Internal.Recognition.CharacterDictionary")!;
        var lines=(IReadOnlyList<string>)dict.GetMethod("LoadLines")!.Invoke(null,new object[]{args[1]})!;
        var recType=sdk.GetType("PaddleOcrNet.Internal.Recognition.SvtrRecognizer")!;
        using var rec=(IDisposable)Activator.CreateInstance(recType,session,lines,48,6)!;
        var normalize=recType.GetMethod("WriteNormalized",BindingFlags.Instance|BindingFlags.NonPublic)!;
        var decode=sdk.GetType("PaddleOcrNet.Internal.Recognition.CtcDecoder")!.GetMethods().Single(m=>m.Name=="GreedyDecode"&&m.GetParameters().Length==5).CreateDelegate<Decode>();
        Console.WriteLine($"PaddleOcrNet={sdk.GetName().Version}; ONNX Runtime={OrtEnv.Instance().GetVersionString()}; {RuntimeInformation.OSDescription}; {RuntimeInformation.ProcessArchitecture}");
        byte[]? previousEffective=null;
        foreach(int width in new[]{279,320}) {
            var input=new DenseTensor<float>(new[]{1,3,48,width});
            // ceil(48 * 180 / 34) = 255. Keep every effective pixel identical.
            normalize.Invoke(rec,new object[]{crop,255,width,input.Buffer,0});
            var effective=new float[3*48*255];
            for(int c=0;c<3;c++)for(int y=0;y<48;y++) {
                input.Buffer.Span.Slice((c*48+y)*width,255).CopyTo(effective.AsSpan((c*48+y)*255,255));
                for(int x=255;x<width;x++)if(input.Buffer.Span[(c*48+y)*width+x]!=0)throw new InvalidDataException("Nonzero padding");
            }
            byte[] bytes=MemoryMarshal.AsBytes(effective.AsSpan()).ToArray();
            if(previousEffective!=null&&!bytes.SequenceEqual(previousEffective))throw new InvalidDataException("Effective pixels changed");
            previousEffective=bytes;
            using var output=session.Run(new[]{NamedOnnxValue.CreateFromTensor(session.InputMetadata.Keys.First(),input)});
            var scores=output.First().AsTensor<float>();int time=scores.Dimensions[1],classes=scores.Dimensions[2];
            var vocab=(IReadOnlyList<string>)dict.GetMethod("BuildVocab")!.Invoke(null,new object[]{lines,classes})!;
            float[] values=scores.ToArray();var text=decode(values,time,classes,vocab,null);
            int space=vocab.ToList().IndexOf(" ");int step=10,best=0;
            for(int c=1;c<classes;c++)if(values[step*classes+c]>values[step*classes+best])best=c;
            Console.WriteLine($"W={width}, effective=255, text=[{text.Text}], step10.argmax={best}, p(space)={values[step*classes+space]:F8}, p(blank)={values[step*classes]:F8}");
        }
    }
}
```
</details>
